### ISSUE TITLE
Feature request: Support secret creation for github actions

### ISSUE THREAD

**Issue by gjambet** [terraform-providers/terraform-provider-github#271]
Hi friends,

 it would be super cool to be able to set secrets for actions through terraform while creating a repository. ;)

ty, 

g. (happy action user)

**Comment by hronix** [comment]
Yes please

**Comment by zzh8829** [comment]
github hasn't made api for secrets available yet, we have to wait for now

**Comment by iniinikoski** [comment]
Hi. Any updates on this - is the Secrets API still private...?

**Comment by gjambet** [comment]
❤️

**Comment by jcudit** [comment]
GitHub is aiming to make this endpoint available within the next few sprints.  Mention me here if I don't post an update in a month or so.

**Comment by ajliv** [comment]
Secrets have been added to the GitHub API https://developer.github.com/v3/actions/secrets/

**Comment by martinssipenko** [comment]
We now need to wait till Go GitHub client releases action secrets - https://github.com/google/go-github/pull/1402

**Comment by benesis002** [comment]
+1

**Comment by btkostner** [comment]
It looks like it was merged https://github.com/google/go-github/pull/1402#issuecomment-583939009 and released in the [v29.0.3 tag](https://github.com/google/go-github/blob/v29.0.3/github/actions_secrets.go).

**Comment by martinssipenko** [comment]
I think we should also wait for #342 before we start implementing this one.

**Comment by martinssipenko** [comment]
What do we think about usage? Should action secrets be a new resource - `github_repository_action_secret` and be used as follows?

```hcl
resource "github_repository_action_secret" "foo" {
    repository = "${github_repository.repo.name}"
    name       = "FOO"
    value      = "secret"
}
```

**Comment by hazcod** [comment]
Is there a way to provide ciphertexts in terraform and supply the key via an environment variable?
I can imagine it's problematic to store plaintext github secrets in terraform files ... which might be located on github again.

**Comment by martinssipenko** [comment]
@hazcod I think one option would be that resource accepts two conflicting arguments - `secret` and `encrypted_secret`, The `secret` could be used to pass plain text value which would be encrypted by terraform using GitHubs public key. The `encrypted_secret` would accept already encrypted value and leave the encryption the user.

Off topic: In order to not store secrets in terraform files I use an approach where I store secrets somewhere else, and then use a data provider to get the secret values from there. In that case I only need to reference the value from data source.

**Comment by martinssipenko** [comment]
@radeksimko @jcudit @paultyng what are your thoughts on this?

**Comment by tgermain** [comment]
Note for the implementation: how to know if a secret has been changed and need to be updated by terraform ?

The [secret API](https://developer.github.com/v3/actions/secrets/#get-a-secret) does not offer a way to read the content of a secret (normal) but return the `created_at` and `updated_at` date. 

Should we store at least the `updated_at` date in terraform state and when we refresh terraform state (before a plan for instance) we get the secret and compare its `update_at` date with the one in the state. 
A difference in those dates would indicate the secret was updated since the last time terraform interact with. What do you think?

**Comment by martinssipenko** [comment]
Another option is that terraform does not detect and repair "drift" on this resource in case the data is updated or deleted outside of Terraform. I think it might also be possible to not store secrets in state, but instead store a hash value of the secret, in case current value hash does not match hash stored in state an update should be performed.

**Comment by StephenWithPH** [comment]
A lurker (who wants this feature) chiming in...

I suggest not trying to get too clever regarding secrets in state. I often treat the Terraform Vault Provider as a guide for the official-ish Hashicorp way of doing things in Terraform.

>Interacting with Vault from Terraform causes any secrets that you read and write to be persisted in both Terraform's state file and in any generated plan files. For any Terraform module that reads or writes Vault secrets, these files should be treated as sensitive and protected accordingly.

(https://www.terraform.io/docs/providers/vault/index.html)

Directly on the label, they disclaim any cleverness to protect secrets.

I think Terraform core offers enough ways to protect secrets. The combination of using remote state and an encryption-enabled backend seems sufficient. See https://www.terraform.io/docs/state/sensitive-data.html.

**Comment by tgermain** [comment]
Indeed @StephenWithPH , state is already sensitive so we do not have to be too clever using hash in state (instead of plaintext values).

> Another option is that terraform does not detect and repair "drift" on this resource in case the data is updated or deleted outside of Terraform.

This is also fine if it is clearly mentioned in the doc of the resource. 
In case of doubts, you can always _taint_ the doubtful resource in your terraform and recreate it. Terraform is your source of truth so manuals changes outside of it should be committed in terraform if they are meant to be permanent.

**Comment by senare** [comment]
Nice to see that this worked!

imho ... just get it done ... KISS

**Comment by benj-fletch** [comment]
Hi, I'd like to take on this as we have a requirement for it in our organisation.
I think that we'd need to merge in https://github.com/terraform-providers/terraform-provider-github/pull/342 before work can be undertaken here, however, since v29 of `go-github` contains the calls we would need to utilise.
If it's ok, I'll start work on this in a branch of master merged with https://github.com/terraform-providers/terraform-provider-github/pull/342 and we can go from there?

**Comment by martinssipenko** [comment]
@benj-fletch if you need any help, feel free to ask me.

**Comment by benj-fletch** [comment]
@martinssipenko Thanks!
I think from reading the comments I'll keep it relatively simple and allow the user of the resource to protect their secrets as they see fit (where possible). I need to see what is available to configure on the API and work on that basis.

**Comment by jcudit** [comment]
@benj-fletch my only ask on this one is to try to keep things forward compatible with these [related enhancements](https://github.com/google/go-github/issues/1399#issuecomment-579554516).  I'm 👍 on designing for overall Actions support rather than just the `secret` functionality.

**Comment by benj-fletch** [comment]
@jcudit Sure thing, I will try my best to do so!

**Comment by techjacker** [comment]
Any idea when this will get merged? Would be very handy feature

**Comment by benj-fletch** [comment]
I believe that this ticket can be closed now, following the merge of #362 ?
